src/integration-tests/src/storage.rs

```diff
@@ -76,6 +76,63 @@ pub async fn objects(builder: storage::client::ClientBuilder) -> Result<()> {
     Ok(())
 }
 
+pub async fn objects_customer_supplied_encryption(
```

Comment:
Oh, I see, you need `insert_object()` with support of encryption so you can test the reading function... Hmm... I think it would be better if that was a separate PR, but the code looks great.

Reply:
Exactly 😄  Moved the basic tests and initial refactor to #2369